RIPS/rip-9996.md

```diff
@@ -0,0 +1,140 @@
+
+Having a more agile ordering and replay protection system will significantly improve the user experience with RIP-7560.
+
+Additionally, this change is required for the Native Account Abstraction to achieve feature parity and compatibility
+with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution described below.
+
+## Specification
+
+```
+ABSTRACT_NONCE_SUBTYPE = x
+```
+
+### Non-sequential nonce support
+
+We propose an introduction of a separate 2-dimensional `nonce` value used when contracts initiate a transaction.
+
+The `nonce` parameter of the transaction is to be interpreted as `uint192 key || uint64 seq` value.
+The contract account nonce is then defined as a mapping `address account => uint192 key => uint64 seq`.
+This approach guarantees unique transaction nonce and hash but removes the requirement of nonces being sequential
+numbers.
+
+This `nonce` is exposed to the EVM in a `NonceManager` pre-deployed contract located at the `AA_NONCE_MANAGER` address.
+
+The `nonce` is [validated and incremented](#nonce-validation-frame) on-chain before the rest of the validation code.
+
+The old `nonce` account parameter remains in use for transactions initiated by EOAs and for the `CREATE` opcode.
+It is not incremented when `sender` makes the `ABSTRACT_NONCE_SUBTYPE` transaction.
+
+#### Nonce validation frame
+
+Before the first validation frame is applied during the validation phase,
+the `NonceManager` is invoked with the following data:
+
+```solidity
+abi.encodePacked(sender, nonce)
+```
+
+The gas costs of this execution frame are counted towards the total for a transaction.
+
+#### NonceManager Pseudocode
+
+```
+
+if evm.caller == AA_ENTRY_POINT:
+    validate_increment()
+else:
+    get()
+
+def get():
+    if len(evm.calldata) != 44:
+        evm.revert()
+
+    // address sender, uint192 key
+    address = to_uint160_be(evm.calldata[0:20])
+    key = to_uint192_be(evm.calldata[20:44])
+
+    nonce = storage.get(keccak(address, key))
+
+    evm.return((key << 64) + nonce)
+
+def validate_increment():
+
+    address = to_uint160_be(evm.calldata[0:20])
+    key = to_uint192_be(evm.calldata[20:44])
+    nonce = to_uint64_be(evm.calldata[44:52])
+
+    current_nonce = storage.get(keccak(address, key))
+
+    if (nonce != current_nonce):
+        evm.revert()
+
+    storage.set(kecca
+    k(address, key), current_nonce + 1)
+
+```
+
+#### NonceManager Bytecode and deployment
+
+TODO.
+
+## Rationale
+
+### Creating a pre-deployed contract instead of modifying an account data structure
+
+While there is no technical benefit to either option, allowing EVM code to control the nonce
+seems to be a smaller change to the Ethereum protocol and is more aligned with the vision of Account Abstraction.
+
+## Backwards Compatibility
+
+As actual `nonce` value was never observable inside the EVM, there should be no major complications caused by the
```

Comment:
I am guessing this would not be technically feasible to expose an existing opcode to a NonceManager contract that is being deployed. There is no precedent for such an arrangement in the EVM, and I don't think it should be done here.